### PR TITLE
Fix #2061 (don't suggest removing types which implement Drop)

### LINKED ISSUE #2061: unnecessary_operation false positive with Drop
On the following code:

```rust
use std::rc::Rc;

struct Foo {
    _r: Rc<u32>,
}

impl Drop for Foo {
    fn drop(&mut self) {
        println!("Foo");
    }
}

fn main() {
    Foo { _r: Rc::new(1) };  // clippy: replace with "Rc::new(1)"
}
```

clippy suggests removing the `Foo{}` wrapper, which would cause `Foo::drop` not to be called:

```
warning: statement can be reduced
  --> src/main.rs:14:5
   |
14 |     Foo { _r: Rc::new(1) };
   |     ^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `Rc::new(1);`
   |
   = note: #[warn(unnecessary_operation)] on by default
```

### PR DESCRIPTION
**Pull request by jugglerchris**
Update unnecessary_operation and no_effect to not suggest removing structs/enums wrappers when that type implements Drop.

It works for all the test cases I added (and doesn't break anything) but I'm not sure about the case for `ExprPath` (for catching unit structs) doesn't have any undesired side effects for example.

The `has_drop` code for checking if a type implements `Drop` was taken from another file; maybe it'd be better in a utility module.

### CODE REVIEW

File: clippy_lints/src/no_effect.rs [hunk 1]
```diff
@@ -40,12 +40,22 @@ declare_lint! {
     "outer expressions with no effect"
 }
 
+/// Check whether this type implements Drop.
+fn has_drop(cx: &LateContext, expr: &Expr) -> bool {
```
Comment (Manishearth) [inline clippy_lints/src/no_effect.rs]: Yeah, please put this somewhere in `utils`

**Comment by oli-obk** [comment]
I would prefer if we had a whitelist of common things that are allowed to have drop but are known to have no side effects. Like strings or vectors of whitelisted things

File: clippy_lints/src/no_effect.rs (hunk 1 above)
Comment (jugglerchris) [reply clippy_lints/src/no_effect.rs]: Sure.

**Comment by jugglerchris** [comment]
@oli-obk While thinking about this I've realised that the things that would be in the whitelist (i.e. some of the types with destructors in the standard library) are generally made via some kind of function call (`Rc::new(...)`) than with a raw struct/enum constructor (`Foo { field: val }`).  The exceptions (given that a destructor probably means you want to control construction too) are probably only really in the implementations of those functions and test code (so inside stdlib code for things that would be on this whitelist).

At the moment (if my understanding is correct) function calls aren't matched in this lint yet, but it could be enhanced to look for things like `std::rc::Rc::new(foo);` and `Vec::new();`.  I haven't yet worked out what `vec![1,2,3]` turns into.

**Comment by oli-obk** [comment]
Right. I didn't think that through. Still, a struct with a string field also implements drop, even if not explicitly. We definitely want to lint useless struct construction even for such structs though. Can you check how your changes treat such a struct?

**Comment by jugglerchris** [comment]
My changes don't seem to affect what happens with a struct with a `String` field:

```
error: statement can be reduced
   --> tests/ui/no_effect.rs:121:5
    |
121 |     FooString { s: String::from("blah"), };
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `String::from("blah");`

error: statement with no effect
   --> tests/ui/no_effect.rs:123:5
    |
123 |     FooString { s: s };
    |     ^^^^^^^^^^^^^^^^^^^

```

**Comment by jugglerchris** [comment]
By the way, is there a way to run a single one of the ui tests?  I couldn't see an obvious way.

**Comment by llogiq** [comment]
prepend `TESTNAME=foo` (where `foo` is the name of your test).

**Comment by jugglerchris** [comment]
@llogiq Thanks!

**Comment by jugglerchris** [comment]
Hi - is there anything needed/wanted on this PR?

**Comment by oli-obk** [comment]
Just resolve the merge conflict and add the test case with the string to the test suite.

Sorry I tottally forgot about the PR

**Comment by jugglerchris** [comment]
No problem! I've added a test and fixed the merge conflict.

**Comment by oli-obk** [comment]
Great! Another bug bites the dust

**Pull request merged**